sdks/java/extensions/scripting/src/main/java/org/apache/beam/sdk/extensions/scripting/ScriptingParDo.java

```diff
@@ -0,0 +1,189 @@
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TypeDescriptor;
+
+/**
+ * A transform to write simple ParDo transformations with a scripting language
+ * supported by Java JSR-223.
+ */
+@Experimental
+public abstract class ScriptingParDo<InputT, OutputT>
+    extends PTransform<PCollection<InputT>, PCollection<OutputT>> {
+
+  private ValueProvider<String> language = StaticValueProvider.of("js");
+  private ValueProvider<String> script = StaticValueProvider.of("");
+  private Coder<OutputT> coder;
+
+  public static <InputT, OutputT> ScriptingParDo<InputT, OutputT> of(final Coder<OutputT> coder) {
+    final ScriptingParDo<InputT, OutputT> scripting = new ScriptingParDo<InputT, OutputT>() {};
+    scripting.coder = coder;
+    return scripting;
+  }
+
+  public ScriptingParDo<InputT, OutputT> withLanguage(final String language) {
+    checkArgument(language != null, "language can not be null");
+    return withLanguage(ValueProvider.StaticValueProvider.of(language));
+  }
+
+  public ScriptingParDo<InputT, OutputT> withLanguage(final ValueProvider<String> language) {
+    checkArgument(language != null, "language can not be null");
+    this.language = language;
+    return this;
+  }
+
+  public ScriptingParDo<InputT, OutputT> withScript(final String script) {
+    checkArgument(script != null, "script can not be null");
+    return withScript(ValueProvider.StaticValueProvider.of(script));
+  }
+
+  public ScriptingParDo<InputT, OutputT> withScript(final ValueProvider<String> script) {
+    checkArgument(script != null, "script can not be null");
+    this.script = script;
+    return this;
+  }
+
+  @Override
+  public PCollection<OutputT> expand(final PCollection<InputT> apCollection) {
+    if (language == null || script == null || script.get().isEmpty()) {
+      throw new IllegalArgumentException("Language and Script must be set");
+    }
+    return apCollection.apply(ParDo.of(new ScriptingDoFn<>(language.get(), script.get())));
+  }
+
+  @Override // ensure we don't always need to set the coder
+  public <T> Coder<T> getDefaultOutputCoder(
+      final PCollection<InputT> input, final PCollection<T> output)
+      throws CannotProvideCoderException {
+    if (coder != null) {
+      return (Coder<T>) coder;
+    }
+    final Type superclass = getClass().getGenericSuperclass();
+    if (ParameterizedType.class.isInstance(superclass)) {
+      final Type type = ParameterizedType.class.cast(superclass).getActualTypeArguments()[1];
+      return (Coder<T>) output.getPipeline().getCoderRegistry().getCoder(TypeDescriptor.of(type));
+    }
+    return (Coder<T>) SerializableCoder.of(Serializable.class);
+  }
+
+  private static class ScriptingDoFn<InputT, OutputT> extends DoFn<InputT, OutputT> {
+    private String language;
+    private String script;
+
+    private volatile ScriptEngine engine;
+    private CompiledScript compiledScript;
```

Comment:
`CompiledScript` is not `Serialisable` so probably it should be `transient`.

Reply:
ok